### ISSUE TITLE
Crash when using self in imp_implementationWithBlock on armeabi-v7a

### ISSUE THREAD

**Issue by rupertdaniel** [gnustep/libobjc2#126]
On armeabi-v7a, when you try and use self in a imp_implementationWithBlock block, the app crashes.

It works correctly for arm64-v8a, x86 and x86_64.

Attached is a patch to BlockImpTest.m that can be used to reproduce the issue.
[libobjc-block-imp-crash.patch.zip](https://github.com/gnustep/libobjc2/files/3681196/libobjc-block-imp-crash.patch.zip)

**Comment by davidchisnall** [comment]
Please can you put the patch inline, rather than in a zip?

**Comment by rupertdaniel** [comment]
@davidchisnall - updated with patch inline and the reported crash line number

**Comment by davidchisnall** [comment]
Thanks.  It sounds as if the self parameter is being corrupted by the block trampoline.  We should probably add an explicit test that it is the correct value (for both the sret and non-sret case).  

Do you know if your libobjc.so contains the ARM or Thumb-2 version of the trampoline?  You can change [this line](https://github.com/gnustep/libobjc2/blob/b1964451e123050c9062eb605894d712a195cb27/block_trampolines.S#L121) to a 1 or 0 to switch between them explicitly.  I think for ARMv7 Thumb-2 should be the default, but looking at the code I think the `mov \arg0, \arg1` line in both has the arguments the wrong way around.  Can you try swapping the order and see if that fixes it for you?

**Comment by rupertdaniel** [comment]
Yeap changing the lines to `mov \arg1, \arg0` seems to fix it!

Should I do a PR with the changes?

**Comment by davidchisnall** [comment]
Yes please!  Can you do a PR with the changes to both ARM code paths and a block used for the sret and non-sret case that checks that self is the expected value.
